---
title: "R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r message=FALSE, warning=FALSE}
library("DESeq2")
library("tximport")
library("rhdf5")
library("EnsDb.Mmusculus.v79")
library("AnnotationHub")
library("BiocParallel")
library("ggplot2")
library("pheatmap")
library("dplyr")
library("ggpubr")
library("reshape2")
#register(MulticoreParam(32))
```

```{r}
edb = EnsDb.Mmusculus.v79
k <- keys(edb, keytype = "TXNAME")
tx2gene = genes(edb, filter = TxIdFilter(k), columns = c("tx_id","gene_id"), return.type="data.frame")
```

```{r}
setwd("C:/Users/NOBEL/Box/Lab Notes/Lab Notes Benjamin/Sestrins/Mouse RNA-Seq")
condition = factor(rep(c("WT", "KO"), each=2, times=3), levels = c("WT","KO"))
day = factor(rep(c("d-0","d-7","d-21"), each=4), levels = c("d-0","d-7","d-21"))
replicate = factor(rep(c("R1","R2"), times=6))

sample_id = paste(day, condition, replicate, sep="-")
sample_path = file.path(".", "mm38", paste("MuSC", sample_id, sep="-"))
sampleTable = data.frame(sample=as.character(sample_id), condition=condition, day=day, replicate=replicate)
sampleTable <- dplyr::mutate(sampleTable, path = sample_path)
# s2c$day = as.factor(s2c$day)'
sampleTable$group = factor(paste(sampleTable$day,sampleTable$condition,sep = '-'))
sampleTable$group = relevel(sampleTable$group, ref="d-0-WT")
sampleTable = sampleTable[-9,]
sampleTable
```

```{r}
files <- file.path(sampleTable$path, "abundance.h5")
txi.kallisto <- tximport(files, type = "kallisto",  tx2gene=tx2gene, ignoreTxVersion = TRUE)
```

```{r}
saveRDS(txi.kallisto, "./nosva_v4/RDS/txi.kallisto.RDS")
```

```{r}
dds <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, design = ~ group)
dds = DESeq(dds)
keep <- rowSums(counts(dds)) >= 1 # only keep genes that have at least one transcript across all samples
dds <- dds[keep,]
resultsNames(dds)
```

```{r}
saveRDS(dds, "./nosva_v4/RDS/raw_dds.RDS")
```

```{r}
dds_cor = cor(counts(dds),  method="spearman")
rownames(dds_cor) <- paste(dds$sample, sep="-")
colnames(dds_cor) <- NULL
pheatmap(dds_cor, main="Spearman Correlation")
```

# Visualization.
```{r}
# Want variance ranges to remain relatively constant across mean values.
# Use regularized-logarithm transformation b/c working with small dataset (n<30).
rld = rlog(dds, blind=FALSE)
head(assay(rld), 3)
```

```{r}
saveRDS(rld, "./nosva_v4/RDS/rld.RDS")
```

```{r}
library("vsn")
tiff(filename="./nosva_v4/Figures/dds_rld.tiff")
dds_rld = meanSdPlot(assay(rld)) # SD looks fairly constant with rld transformation
# Note that the vertical axis in such plots is the square root of the variance over all samples, so including the variance due to the experimental conditions. While a flat curve of the square root of variance over the mean may seem like the goal of such transformations, this may be unreasonable in the case of datasets with many true differences due to the experimental conditions.
dev.off()
```
 
```{r}
sampleDists <- dist(t(assay(rld)))
sampleDists
library("pheatmap")
library("RColorBrewer")
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(rld$sample, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         main="Distance Matrix",
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         color = colors)

# A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.
```

```{r}
pcaData <- plotPCA(rld, intgroup = c( "condition", "day"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# dds_pca = ggplot(pcaData, aes(x = PC1, y = PC2, color = condition, shape = day)) +
#       geom_point(size=5) +
#       xlab(paste0("PC1: ", percentVar[1], "% variance")) +
#       ylab(paste0("PC2: ", percentVar[2], "% variance")) +
#   theme_pubr()
dds_pca = ggscatter(pcaData, x="PC1", y="PC2", 
                    color="condition", shape="day",
                    palette = "lancet",
                    size=5,
                    legend="right") + 
  labs_pubr()
dds_pca

ggsave(dds_pca, filename="./nosva_v4/Figures/rlog_pca.tiff", 
       width=5, height=4, units="in", dpi=320)
```

```{r}
# Heatmap of count matrix
select <- order(rowMeans(counts(dds,normalized=TRUE)), decreasing=TRUE)[1:100]
df <- as.data.frame(colData(dds)[,c("condition","day")])

count_hmp = assay(rld)[select,]
rownames(count_hmp) = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(count_hmp)), return.type="data.frame")$symbol
#colnames(count_hmp) = paste(df$condition, df$day, c("R1","R2"), sep="-")
tiff(filename="./nosva_v4/Figures/countmatrix_pheatmap.tiff", res=100)
pheatmap(count_hmp, cluster_rows=TRUE, show_rownames=TRUE, scale="row", cellwidth = 15,
         cluster_cols=TRUE, annotation_col=df)

dev.off()
```

## Create LFC shrunken results objects for KO vs WT each day
```{r}
res_d0 = results(dds, name="group_d.0.KO_vs_d.0.WT", pAdjustMethod = "fdr")
resLFC_d0 = lfcShrink(dds, res=res_d0, coef="group_d.0.KO_vs_d.0.WT", type="ashr")
resLFC_d0$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_d0)), return.type="data.frame")$symbol
resLFC_d0 = resLFC_d0[order(resLFC_d0$padj),]

res_d7 = results(dds, contrast=c("group","d-7-KO","d-7-WT"), pAdjustMethod = "fdr")
resLFC_d7 = lfcShrink(dds, res=res_d7, contrast=c("group","d-7-KO","d-7-WT"), type="ashr")
resLFC_d7$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_d7)), return.type="data.frame")$symbol
resLFC_d7 = resLFC_d7[order(resLFC_d7$padj),]

res_d21 = results(dds, contrast=c("group","d-21-KO","d-21-WT"), pAdjustMethod = "fdr")
resLFC_d21 = lfcShrink(dds, res=res_d21, contrast=c("group","d-21-KO","d-21-WT"), type="ashr")
resLFC_d21$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_d21)), return.type="data.frame")$symbol
resLFC_d21 = resLFC_d21[order(resLFC_d21$padj),]

summary(resLFC_d0)
summary(resLFC_d7)
summary(resLFC_d21)
```

```{r}
saveRDS(resLFC_d0, "./nosva_v4/RDS/resLFC_d0.RDS")
saveRDS(resLFC_d7, "./nosva_v4/RDS/resLFC_d7.RDS")
saveRDS(resLFC_d21, "./nosva_v4/RDS/resLFC_d21.RDS")
```

# Day 0 different significance levels
```{r}
resLFC_d0_Down_Sig0.1 = subset(resLFC_d0, subset=(padj<=0.1 & log2FoldChange<=-1))
resLFC_d0_Up_Sig0.1 = subset(resLFC_d0, subset=(padj<=0.1 & log2FoldChange>=1))
summary(resLFC_d0_Down_Sig0.1)
summary(resLFC_d0_Up_Sig0.1)

resLFC_d0_Down_Sig0.05 = subset(resLFC_d0, subset=(padj<=0.05 & log2FoldChange<=-1))
resLFC_d0_Up_Sig0.05 = subset(resLFC_d0, subset=(padj<=0.05 & log2FoldChange>=1))
summary(resLFC_d0_Down_Sig0.05)
summary(resLFC_d0_Up_Sig0.05)

resLFC_d0_Down_Sig0.01 = subset(resLFC_d0, subset=(padj<=0.01 & log2FoldChange<=-1))
resLFC_d0_Up_Sig0.01 = subset(resLFC_d0, subset=(padj<=0.01 & log2FoldChange>=1))
summary(resLFC_d0_Down_Sig0.01)
summary(resLFC_d0_Up_Sig0.01)

write.csv(resLFC_d0_Down_Sig0.1, file="./nosva_v4/d0_sig0.1_down_l2fc1.csv")
write.csv(resLFC_d0_Up_Sig0.1, file="./nosva_v4/d0_sig0.1_up_l2fc1.csv")
write.csv(resLFC_d0_Down_Sig0.05, file="./nosva_v4/d0_sig0.05_down_l2fc1.csv")
write.csv(resLFC_d0_Up_Sig0.05, file="./nosva_v4/d0_sig0.05_up_l2fc1.csv")
write.csv(resLFC_d0_Down_Sig0.01, file="./nosva_v4/d0_sig0.01_down_l2fc1.csv")
write.csv(resLFC_d0_Up_Sig0.01, file="./nosva_v4/d0_sig0.01_up_l2fc1.csv")
```

# Day 7 different significance levels
```{r}
resLFC_d7_Down_Sig0.1 = subset(resLFC_d7, subset=(padj<=0.1 & log2FoldChange<=-1))
resLFC_d7_Up_Sig0.1 = subset(resLFC_d7, subset=(padj<=0.1 & log2FoldChange>=1))
summary(resLFC_d7_Down_Sig0.1)
summary(resLFC_d7_Up_Sig0.1)

resLFC_d7_Down_Sig0.05 = subset(resLFC_d7, subset=(padj<=0.05 & log2FoldChange<=-1))
resLFC_d7_Up_Sig0.05 = subset(resLFC_d7, subset=(padj<=0.05 & log2FoldChange>=1))
summary(resLFC_d7_Down_Sig0.05)
summary(resLFC_d7_Up_Sig0.05)

resLFC_d7_Down_Sig0.01 = subset(resLFC_d7, subset=(padj<=0.01 & log2FoldChange<=-1))
resLFC_d7_Up_Sig0.01 = subset(resLFC_d7, subset=(padj<=0.01 & log2FoldChange>=1))
summary(resLFC_d7_Down_Sig0.01)
summary(resLFC_d7_Up_Sig0.01)

write.csv(resLFC_d7_Down_Sig0.1, file="./nosva_v4/d7_sig0.1_down_l2fc1.csv")
write.csv(resLFC_d7_Up_Sig0.1, file="./nosva_v4/d7_sig0.1_up_l2fc1.csv")
write.csv(resLFC_d7_Down_Sig0.05, file="./nosva_v4/d7_sig0.05_down_l2fc1.csv")
write.csv(resLFC_d7_Up_Sig0.05, file="./nosva_v4/d7_sig0.05_up_l2fc1.csv")
write.csv(resLFC_d7_Down_Sig0.01, file="./nosva_v4/d7_sig0.01_down_l2fc1.csv")
write.csv(resLFC_d7_Up_Sig0.01, file="./nosva_v4/d7_sig0.01_up_l2fc1.csv")
```

# Day 21 different significance levels
```{r}
resLFC_d21_Down_Sig0.1 = subset(resLFC_d21, subset=(padj<=0.1 & log2FoldChange<=-1))
resLFC_d21_Up_Sig0.1 = subset(resLFC_d21, subset=(padj<=0.1 & log2FoldChange>=1))
summary(resLFC_d21_Down_Sig0.1)
summary(resLFC_d21_Up_Sig0.1)

resLFC_d21_Down_Sig0.05 = subset(resLFC_d21, subset=(padj<=0.05 & log2FoldChange<=-1))
resLFC_d21_Up_Sig0.05 = subset(resLFC_d21, subset=(padj<=0.05 & log2FoldChange>=1))
summary(resLFC_d21_Down_Sig0.05)
summary(resLFC_d21_Up_Sig0.05)

resLFC_d21_Down_Sig0.01 = subset(resLFC_d21, subset=(padj<=0.01 & log2FoldChange<=-1))
resLFC_d21_Up_Sig0.01 = subset(resLFC_d21, subset=(padj<=0.01 & log2FoldChange>=1))
summary(resLFC_d21_Down_Sig0.01)
summary(resLFC_d21_Up_Sig0.01)

write.csv(resLFC_d21_Down_Sig0.1, file="./nosva_v4/d21_sig0.1_down_l2fc1.csv")
write.csv(resLFC_d21_Up_Sig0.1, file="./nosva_v4/d21_sig0.1_up_l2fc1.csv")
write.csv(resLFC_d21_Down_Sig0.05, file="./nosva_v4/d21_sig0.05_down_l2fc1.csv")
write.csv(resLFC_d21_Up_Sig0.05, file="./nosva_v4/d21_sig0.05_up_l2fc1.csv")
write.csv(resLFC_d21_Down_Sig0.01, file="./nosva_v4/d21_sig0.01_down_l2fc1.csv")
write.csv(resLFC_d21_Up_Sig0.01, file="./nosva_v4/d21_sig0.01_up_l2fc1.csv")
```

```{r dpi=200}
library("genefilter")
library("pheatmap")
library("RColorBrewer")
library(ComplexHeatmap)

#topVarGenes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 50)
topVarGenes <- unlist(lapply(unique(unlist(d0_d7_d21_int)), FUN = function(x) {match(x,rownames(assay(rld)))}))

mat  <- assay(rld)[ topVarGenes, ]
#mat  <- mat - rowMeans(mat)
topVarGeneSymbols = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(mat)), return.type="data.frame")$symbol
rownames(mat) = topVarGeneSymbols
anno_col <- as.data.frame(colData(rld)[, c("condition","day")])
anno_col

dpgp_clust = read.table("./nosva_v4/DPGP/DESeq2/dpgp_clusters.tsv", sep="\t", as.is = T, header=T, row.names = NULL)
anno_row <- data.frame(cluster=factor(
  dpgp_clust$cluster[
    unlist(lapply(rownames(mat), function(x) {match(x,dpgp_clust$symbol)}))
    ]), stringsAsFactors = FALSE)
anno_row = cbind(mat[,0],anno_row)
anno_row
# 
# row_ha = rowAnnotation(cluster = anno_block(gp = gpar(fill = brewer.pal(11,"Paired")[order()])))
# Heatmap(t(scale(t(mat))), cluster_rows = T, cluster_columns = F, right_annotation = row_ha,
#         row_km=11)

# hmp = pheatmap(mat, annotation_col = anno, scale="row", 
#                color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
#                show_colnames = FALSE, show_rownames = TRUE, fontsize_row = 8,
#                cellwidth = 40, cellheight = 9, border_color = NA,
#                cluster_rows = TRUE, cluster_cols = FALSE, treeheight_row = 0,
#                annotation_legend = FALSE, legend = TRUE) 

hmp_dpgp = pheatmap(mat, annotation_col = anno_col, annotation_row = anno_row, cluster_rows = F, scale="row",
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
         show_rownames = F, cellwidth = 30, cellheight = 0.4, cluster_cols = F, show_colnames = F)
ggsave(hmp_dpgp, filename="./nosva_v4/Figures/DE_rld_KOvsWT_DPGPclust_pheatmap.tiff", units="in", width=7, height=6.5, dpi=200, limitsize=FALSE)

hmp = pheatmap(mat, annotation_col = anno_col, annotation_row = anno_row, scale="row", 
               color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
               show_colnames = FALSE, show_rownames = FALSE, fontsize_row = 8,
               cellwidth = 30, cellheight=0.4, border_color = NA,
               cluster_rows = TRUE, cluster_cols = FALSE, treeheight_row = 0,
               annotation_legend = TRUE, legend = TRUE) 
hmp

#ggsave(hmp, filename="./nosva_v4/Figures/DE_rld_KOvsWT_pheatmap.tiff", units="in", width=7, height=6.5, dpi=200, limitsize = FALSE)
```

```{r}
head(rownames(mat[hmp$tree_row[["order"]],]))
hmp_clust = sort(cutree(hmp$tree_row, k=11))
hmp_clust = as.data.frame(hmp_clust, row.names=names(hmp_clust))
hmp_clust$dpgp_clust = hmp_clust[rownames(anno_row),]
hmp_clust
```

```{r fig.height=8}
library("EnhancedVolcano")
# The default cut-off for log2FC is >|2|; the default cut-off for P value is 10e-6.
EnhancedVolcano(resLFC_d0,
  lab = resLFC_d0$Symbol,
  title = "Day 0 (KO vs WT) - 0.01 FDR",
  ylab = bquote(~-Log[10]~adjusted~italic(P)),
  pCutoff = 0.01,
  FCcutoff = 1,
  x = 'log2FoldChange',
  y = 'padj')
EnhancedVolcano(resLFC_d7,
  lab = resLFC_d7$Symbol,
  title = "Day 7 (KO vs WT) - 0.01 FDR",
  ylab = bquote(~-Log[10]~adjusted~italic(P)),
  pCutoff = 0.05,
  x = 'log2FoldChange',
  y = 'padj')
EnhancedVolcano(resLFC_d21,
  lab = resLFC_d21$Symbol,
  title = "Day 21 (KO vs WT) - 0.01 FDR",
  ylab = bquote(~-Log[10]~adjusted~italic(P)),
  pCutoff = 0.01,
  x = 'log2FoldChange',
  y = 'padj')
```

```{r}
plotMA(resLFC_d0, ylim=c(-12,12), main="Day 0 (KO vs WT) - 0.1 FDR")
plotMA(resLFC_d7, ylim=c(-12,12), main="Day 7 (KO vs WT) - 0.05 FDR")
plotMA(resLFC_d21, ylim=c(-12,12), main="Day 21 (KO vs WT) - 0.01 FDR")
```

# Venn Diagram of each day's DE genes
```{r}
library(ggVennDiagram)
d0_d7_d21_int = list(d0=rownames(subset(resLFC_d0, subset=(padj<=0.01 & abs(log2FoldChange)>=1))),
                     d7=rownames(subset(resLFC_d7, subset=(padj<=0.01 & abs(log2FoldChange)>=1))),
                     d21=rownames(subset(resLFC_d21, subset=(padj<=0.01 & abs(log2FoldChange)>=1))))
d0_d7_d21_venn = ggVennDiagram(d0_d7_d21_int, lty=1, lwd=1, color="black") + theme(text=element_text(size=16, family="sans"))
d0_d7_d21_venn_nolabel = ggVennDiagram(d0_d7_d21_int, lty=1, lwd=0.75, color="black", label=NULL) + 
  theme(text=element_text(size=16, family="sans"))
d0_d7_d21_venn
d0_d7_d21_venn_nolabel
ggexport(d0_d7_d21_venn, filename="./nosva_v4/Figures/d0_d7_d21_DE_venn.tiff", dpi=320)
ggexport(d0_d7_d21_venn_nolabel, filename="./nosva_v4/Figures/d0_d7_d21_DE_venn_nolabel.tiff", dpi=320)
```

```{r}
d0_d7_d21_int_mat = make_comb_mat(d0_d7_d21_int)
set_size(d0_d7_d21_int_mat)
UpSet(d0_d7_d21_int_mat)
```

```{r}
library(VennDiagram)

library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(d0_d7_d21_int,
             category.names = c("d0", "d7", "d21"),
             imagetype = "tiff", resolution = 320,
             filename="./nosva_v4/Figures/d0_d7_d21_DE_venn.tiff", output=TRUE,
             
             # Circles
            lwd = 2,
            lty = 'blank',
            fill = myCol,
            
            # Numbers
            cex = 2,
            fontface = "bold",
            fontfamily = "sans",
            
            # Set names
        cat.cex = 2,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.035, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)

```

## Create LFC shrunken results objects for KO vs d.0.WT for each day
```{r}
res_d7_d0 = results(dds, name = "group_d.7.KO_vs_d.0.WT", pAdjustMethod = "fdr")
resLFC_d7_d0 = lfcShrink(dds, res=res_d7_d0, coef = "group_d.7.KO_vs_d.0.WT", type="ashr")
resLFC_d7_d0$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_d7_d0)), return.type="data.frame")$symbol
resLFC_d7_d0 = resLFC_d7_d0[order(resLFC_d7_d0$padj),]

res_d21_d0 = results(dds, name = "group_d.21.KO_vs_d.0.WT", pAdjustMethod = "fdr")
resLFC_d21_d0 = lfcShrink(dds, res=res_d21_d0, coef = "group_d.21.KO_vs_d.0.WT", type="ashr")
resLFC_d21_d0$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_d21_d0)), return.type="data.frame")$symbol
resLFC_d21_d0 = resLFC_d21_d0[order(resLFC_d21_d0$padj),]

summary(resLFC_d7_d0)
summary(resLFC_d21_d0)
```

# Venn Diagram of KO vs d.0.WT for each day
```{r}
d0_d7_d21_int_d0WT = list(d0=rownames(subset(resLFC_d0, subset=(padj<=0.01 & abs(log2FoldChange)>=1))),
                     d7=rownames(subset(resLFC_d7_d0, subset=(padj<=0.01 & abs(log2FoldChange)>=1))),
                     d21=rownames(subset(resLFC_d21_d0, subset=(padj<=0.01 & abs(log2FoldChange)>=1))))
d0_d7_d21_int_d0WT_up = list(d0=rownames(subset(resLFC_d0, subset=(padj<=0.01 & log2FoldChange>=1))),
                     d7=rownames(subset(resLFC_d7_d0, subset=(padj<=0.01 & log2FoldChange>=1))),
                     d21=rownames(subset(resLFC_d21_d0, subset=(padj<=0.01 & log2FoldChange>=1))))
d0_d7_d21_int_d0WT_down = list(d0=rownames(subset(resLFC_d0, subset=(padj<=0.01 & log2FoldChange<=-1))),
                     d7=rownames(subset(resLFC_d7_d0, subset=(padj<=0.01 & log2FoldChange<=-1))),
                     d21=rownames(subset(resLFC_d21_d0, subset=(padj<=0.01 & log2FoldChange<=-1))))

d0_d7_d21_venn = ggVennDiagram(d0_d7_d21_int_d0WT, lty=1, lwd=0.75, color="black", n.sides=5e4) + theme(text=element_text(size=16, family="sans"))
d0_d7_d21_venn_up = ggVennDiagram(d0_d7_d21_int_d0WT_up, lty=1, lwd=0.75, color="black", n.sides=5e4) + theme(text=element_text(size=16, family="sans"))
d0_d7_d21_venn_down = ggVennDiagram(d0_d7_d21_int_d0WT_down, lty=1, lwd=0.75, color="black", n.sides=5e4) + theme(text=element_text(size=16, family="sans"))

d0_d7_d21_venn
d0_d7_d21_venn_up
d0_d7_d21_venn_down

ggexport(d0_d7_d21_venn, filename="./nosva_v4/Figures/d0_d7_d21_DE_vsd0WT_venn.tiff", dpi=320)
ggexport(d0_d7_d21_venn_up, filename="./nosva_v4/Figures/d0_d7_d21_DE_vsd0WT_venn_up.tiff", dpi=320)
ggexport(d0_d7_d21_venn_down, filename="./nosva_v4/Figures/d0_d7_d21_DE_vsd0WT_venn_down.tiff", dpi=320)
```

```{r}
library(VennDiagram)
venn_d0WT_up = get.venn.partitions(x = d0_d7_d21_int_d0WT_up, keep.elements = TRUE)
venn_d0WT_down = get.venn.partitions(x = d0_d7_d21_int_d0WT_down, keep.elements = TRUE)
head(venn_d0WT_up)
write.csv(venn_d0WT_up$..values..$`1`, file="./nosva_v4/d0_d7_d21_int_d0WT_up.csv")
write.csv(venn_d0WT_down$..values..$`1`, file="./nosva_v4/d0_d7_d21_int_d0WT_down.csv")
```

# Gene Lists
```{r}
ROS_genes = c("Sesn1", "Sesn2", "Sesn3", 
              "Foxo1", "Foxo3", "Foxo4", "Foxo6",
              "Nfe2l2", "Gadd45a", "Prkaa1", "Prkaa2", "Prx", "Srxn1", 
              "Keap1")
# ROS_genes_ens = genes(edb, columns="symbol", filter=SymbolFilter(ROS_genes), return.type="data.frame")$gene_id
# res_d0[ROS_genes_ens,]
ROS_genes = c("ENSMUSG00000038332","ENSMUSG00000028893","ENSMUSG00000032009", "ENSMUSG00000026773",
              "ENSMUSG00000044167", "ENSMUSG00000048756", "ENSMUSG00000042903", "ENSMUSG00000052135",
              "ENSMUSG00000015839", "ENSMUSG00000036390", "ENSMUSG00000050697", "ENSMUSG00000028518",
              "ENSMUSG00000032802", "ENSMUSG00000053198", "ENSMUSG00000003308", "ENSMUSG00000028991",
              "ENSMUSG00000029167", "ENSMUSG00000029512", "ENSMUSG00000004798", "ENSMUSG00000059552",
              "ENSMUSG00000037211", "ENSMUSG00000056501", "ENSMUSG00000022146", "ENSMUSG00000058755",
              "ENSMUSG00000028736", "ENSMUSG00000026459", "ENSMUSG00000009214", "ENSMUSG00000079471",
              "ENSMUSG00000020908", "ENSMUSG00000033196", "ENSMUSG00000056328", "ENSMUSG00000055775",
              "ENSMUSG00000057003", "ENSMUSG00000028991", "ENSMUSG00000020516", "ENSMUSG00000025583",
              "ENSMUSG00000050310")
resLFC_d0[ROS_genes,]
resLFC_d0[ROS_genes,]$Symbol
```

```{r}
ROS_df = subset(as.data.frame(resLFC_d0[ROS_genes,]), select=c("Symbol", "log2FoldChange", "padj", "lfcSE"))
ROS_df$log10padj = -log10(ROS_df$padj)
ROS_df$error_up = ROS_df$log2FoldChange + ROS_df$lfcSE
ROS_df$error_down = ROS_df$log2FoldChange - ROS_df$lfcSE
ROS_df = subset(ROS_df, subset = padj<0.1)
ROS_df

ROS_bar = ggbarplot(ROS_df, x="Symbol", y="log2FoldChange",
                    fill = "log10padj", color = "white",
                    sort.val = "asc", sort.by.groups = TRUE) + 
  gradient_fill("RdBu") +
  coord_flip() +
  scale_x_discrete(position = "left") +
  labs_pubr() +
  geom_errorbar(data=ROS_df, aes(x=Symbol, ymin=error_down, ymax=error_up), width=0.4, size=1)

ROS_bar = ggpar(ROS_bar, legend = c(0.15,0.80), legend.title = "-log10(padj)", font.legend = c(14),
                font.xtickslab = 14, font.ytickslab = 14, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)

ROS_bar

ggexport(ROS_bar, filename="./nosva_v4/Figures/resLFCd0_ROS_bar.tiff", width=800, height=600, res=150)
```

# Gene Lists
```{r}
library(readxl)
#Qui_genes = c("Pax7", "Zbtb20", "Calcr", "Sirt1", "Notch1", "Notch2", "Itm2a", "Cd34") # Quiescence
total_geneList = as.list(read_xlsx(path = "./TargetedGeneLists.xlsx", col_names=TRUE, col_types = "text"))
names(total_geneList) = c("Qui", "CC", "Infl", "Neur", "Adh", "Myo", "Chr", "Wnt", "ROS")
total_geneList = lapply(total_geneList, function(x) x[!is.na(x)])

# NP_genes = c("Ccna2", "Ccnb1", "Ccnb2", "Ccne1") # Negative regulation of proliferation
# Infl_genes = c("Tnfrsf12a", "Irf7", "Cxcl12", "IL10", "Fas", "Ccl3", "C1ra", "C1s1", "C3")
# Neur_genes = c("Ntrk3", "Cadm1", "Nrn1")
# Adh_genes = c("RET", "Itgb1", "Itga6", "Col5a3", "Lamb1", "Col15a1")
# Myo_genes = c("Myf5", "Myog", "Runx1","Myh1","Myh4", "Myh3","Tnnt2", "Myomaker", "Tmem8c")
# Chr_genes = c("Top2a", "Ezh2", "Cbx5", "Myod1", "Setdb1")
# Wnt_genes = c("Camk2g", "Sphk1", "Mef2c", "Sfrp1", "Sfrp")
# ROS_genes = c("Foxo1", "Foxo3", "Foxo4", "Foxo6", "Nfe2l2", "Gadd45a", "Prkaa1", "Prkaa2", "Prx", 
#               "Srxn1", "Keap1", "Mtor", "Rptor", "Ulk1", "Ulk2","Trp53","Rps6kb1","Ppargc1a","Osmr")
# total_geneList = list(Qui=gene_xlsx$Quiescence, CC=gene_xlsx$Cell_Cycle, Infl=gene_xlsx$Inflammation, Neur=gene_xlsx$NeurDev, 
#                       Adh=gene_xlsx$Adhesion, Myo=gene_xlsx$Myogenesis, Chr=gene_xlsx$Chromatin, Wnt=gene_xlsx$Wnt, ROS=gene_xlsx$ROS)

total_geneList_stack = as.data.frame(stack(total_geneList))

total_geneList_ens = genes(edb, columns="gene_id", filter=SymbolFilter(unlist(total_geneList)), return.type="data.frame")$gene_id
gene_df = resLFC_d0[total_geneList_ens,]
gene_df$Cluster = total_geneList_stack$ind[match(gene_df$Symbol, total_geneList_stack$values)]

gene_df = subset(as.data.frame(gene_df), select=c("Symbol", "log2FoldChange", "padj", "lfcSE", "Cluster"))
gene_df$log10padj = -log10(gene_df$padj)
# gene_df = subset(gene_df, subset = padj<0.1)
gene_df = gene_df[order(gene_df$Cluster),]
gene_df
```

```{r}
gene_bar = ggbarplot(as.data.frame(gene_df), x="Symbol", y="log2FoldChange",
                    fill = "log10padj", color = "white",
                    facet.by="Cluster", sort.val = "asc", sort.by.groups = TRUE,
                    merge = TRUE) + 
  gradient_fill("RdBu") +
  coord_flip() +
  scale_x_discrete(position = "left") +
  theme_pubclean() +
  geom_errorbar(data=gene_df, aes(x=Symbol, ymin=log2FoldChange-lfcSE, ymax=log2FoldChange+lfcSE), width=0.4, size=1)

gene_bar = ggpar(gene_bar, legend.title = "-log10(padj)", font.legend = c(14),
                font.xtickslab = 14, font.ytickslab = 10, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)

gene_bar

ggexport(gene_bar, filename="./nosva_v4/Figures/geneList_d0.tiff", width=1000, height=1000, res=150)
```

# Compare gene lists across time points
```{r}
d0_geneList = subset(as.data.frame(resLFC_d0[total_geneList_ens,]), select = c("Symbol", "log2FoldChange", "padj", "lfcSE"))
d0_geneList$Cluster = total_geneList_stack$ind[match(d0_geneList$Symbol, total_geneList_stack$values)]
d0_geneList$day = rep(factor("d0"), nrow(d0_geneList))
d0_geneList$log10padj = -log10(d0_geneList$padj)
#d0_geneList = subset(d0_geneList, subset = padj<0.1)

d7_geneList = subset(as.data.frame(resLFC_d7[total_geneList_ens,]), select = c("Symbol", "log2FoldChange", "padj", "lfcSE"))
d7_geneList$Cluster = total_geneList_stack$ind[match(d7_geneList$Symbol, total_geneList_stack$values)]
d7_geneList$day = rep(factor("d7"), nrow(d7_geneList))
d7_geneList$log10padj = -log10(d7_geneList$padj)
#d7_geneList = subset(d7_geneList, subset = padj<0.1)

d21_geneList = subset(as.data.frame(resLFC_d21[total_geneList_ens,]), select = c("Symbol", "log2FoldChange", "padj", "lfcSE"))
d21_geneList$Cluster = total_geneList_stack$ind[match(d21_geneList$Symbol, total_geneList_stack$values)]
d21_geneList$day = rep(factor("d21"), nrow(d21_geneList))
d21_geneList$log10padj = -log10(d21_geneList$padj)
#d21_geneList = subset(d21_geneList, subset = padj<0.1)

total_df = rbind(d0_geneList, d7_geneList, d21_geneList)
total_df = subset(total_df, subset = padj<0.05)
```

```{r}
Qui_d0_subset = subset(total_df, subset = (Cluster=="Qui" & day=="d0"))
Qui_d0_bar = ggbarplot(Qui_d0_subset, x="Symbol", y="log2FoldChange",
                    fill = "log10padj", color = "white",
                    sort.val = "asc", sort.by.groups = TRUE) + 
  gradient_fill("RdBu") +
  coord_flip() +
  scale_x_discrete(position = "left") +
  labs_pubr() +
  geom_errorbar(data=Qui_d0_subset, aes(x=Symbol, ymin=log2FoldChange-lfcSE, ymax=log2FoldChange+lfcSE), width=0.4, size=1)
  

Qui_d0_bar = ggpar(Qui_d0_bar, legend = c(0.15,0.80), legend.title = "-log10(padj)", font.legend = c(14),
                font.xtickslab = 14, font.ytickslab = 12, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)

Qui_d0_bar

ggexport(Qui_d0_bar, filename="./nosva_v4/Figures/Qui_d0_bar.tiff", width=500, height=700, res=100)
```

```{r}
Qui_barplot_list = list()
# colfunc <- colorRampPalette(c("#d6604d","#f7f7f7", "#4393c3"))
# plot(rep(1,10),col=colfunc(10),pch=19,cex=3)
# low="#67001F", high="#053061",
for(i in 1:length(levels(total_df$day))){
  Qui_subset = subset(total_df, subset = (Cluster=="Qui" & day==levels(total_df$day)[i]))
  p = ggbarplot(Qui_subset, x="Symbol", y="log2FoldChange",
                fill = "log10padj", color="white",
                sort.val = "asc", sort.by.groups = TRUE) +
    scale_fill_gradientn(colours=brewer.pal(n=11, name="RdBu"), limits = c(0,18), breaks = seq(0,18,4)) +
    coord_flip() +
    scale_x_discrete(position = "left") +
    scale_y_discrete(limits=c(-2,1.5), breaks=seq(-2,2,length.out=5)) +
    geom_errorbar(data=Qui_subset, 
                  aes(x=Symbol, ymin=log2FoldChange-lfcSE, ymax=log2FoldChange+lfcSE), 
                  width=0.4, size=1) +
    labs_pubr()
  p = ggpar(p, legend = c(0.15,0.80), legend.title = "-log10(padj)", font.legend = c(12),
                font.xtickslab = 14, font.ytickslab = 12, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)
  Qui_barplot_list[[i]] = p
}

ggarrange(plotlist = Qui_barplot_list, ncol=3, nrow=1, common.legend=TRUE)
```

```{r}
Myo_barplot_list = list()
# colfunc <- colorRampPalette(c("#d6604d","#f7f7f7", "#4393c3"))
# plot(rep(1,10),col=colfunc(10),pch=19,cex=3)
# low="#67001F", high="#053061",
for(i in 1:length(levels(total_df$day))){
  Myo_subset = subset(total_df, subset = (Cluster=="Myo" & day==levels(total_df$day)[i]))
  p = ggbarplot(Myo_subset, x="Symbol", y="log2FoldChange",
                fill = "log10padj", color="white",
                sort.val = "asc", sort.by.groups = TRUE) +
    scale_fill_gradientn(colours=brewer.pal(n=11, name="RdBu"), limits = c(0,10)) +
    coord_flip() +
    scale_x_discrete(position = "left") +
    scale_y_discrete(limits=c(-2,1.5), breaks=seq(-2,2,length.out=5)) +
    geom_errorbar(data=Myo_subset, 
                  aes(x=Symbol, ymin=log2FoldChange-lfcSE, ymax=log2FoldChange+lfcSE), 
                  width=0.4, size=1) +
    labs_pubr()
  p = ggpar(p, legend = c(0.15,0.80), legend.title = "-log10(padj)", font.legend = c(12),
                font.xtickslab = 14, font.ytickslab = 12, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)
  Myo_barplot_list[[i]] = p
}

ggarrange(plotlist = Myo_barplot_list,common.legend=TRUE)
```

# Term "Scores"
```{r}
Qui_names = rownames(subset(total_df, subset=(Cluster=="Qui")))
counts(dds, normalized=TRUE)[which(rownames(counts(dds, normalized=TRUE))==Qui_names),]
aggregate(log2FoldChange~day, data=subset(total_df, subset=(Cluster=="Qui"), select=c("day","log2FoldChange")), FUN=median)
aggregate(log2FoldChange~day, data=subset(total_df, subset=(Cluster=="CC"), select=c("day","log2FoldChange")), FUN=sum)
```

```{r}
gene_line = ggline(total_df, x="day",y="log2FoldChange",
       add=c("mean","boxplot","jitter"), size=1, color="Cluster",
       linetype="solid", palette="npg",
       point.size = 0.5, point.color = "black") +
  labs_pubr() +  geom_hline(yintercept=0, linetype=2)
gene_line = facet(gene_line, facet.by="Cluster", nrow=4)
gene_line = ggpar(gene_line, legend="none")
gene_line
ggexport(gene_line, filename="./nosva_v4/Figures/gene_line.tiff", width=400,height=400, dpi=320)
```

# Plot counts over time
```{r}
#myo_gene_counts = counts(dds,normalized=TRUE)[rownames(subset(gene_df, subset=Cluster=="Myo")),]
myo_gene_counts = assay(rld)[rownames(subset(gene_df, subset=Cluster=="Myo")),]
#myo_gene_counts = rbind(myo_gene_counts, assay(rld)["ENSMUSG00000079471",])
#rownames(myo_gene_counts)[length(rownames(myo_gene_counts))] = "ENSMUSG00000079471"
colnames(myo_gene_counts) = sampleTable$sample

myo_gene_counts_stack = stack(myo_gene_counts)
colnames(myo_gene_counts_stack) = c("geneid", "sample", "counts")
myo_gene_counts_stack = as.data.frame(myo_gene_counts_stack)
myo_gene_counts_stack$day = unlist(lapply(myo_gene_counts_stack$sample, 
                                   FUN = function(x) {unlist(paste0(strsplit(as.character(x),split="-")[[1]][1],
                                                                    strsplit(as.character(x),split="-")[[1]][2]))}))
myo_gene_counts_stack$condition = unlist(lapply(myo_gene_counts_stack$sample, 
                                                FUN = function(x) {strsplit(as.character(x), split="-")[[1]][3]}))
myo_gene_counts_stack$symbol = as.factor(genes(edb, columns="symbol",
                                               filter=GeneIdFilter(myo_gene_counts_stack$geneid[1:(nrow(myo_gene_counts_stack)-1)]), 
                                               return.type="data.frame")$symbol)
levels(myo_gene_counts_stack$symbol) = c("Myod1", "Myog", "Myf5", "Mef2c", "Myh1", "Myh3", "Myh4", "Tmem8c", "Gm7325")
myo_gene_counts_stack

myo_gene_line = ggline(myo_gene_counts_stack, x="day", y="counts", facet.by = "symbol",
                       color="condition", add=c("mean","boxplot","jitter"), 
                       size=1, palette="lancet",
                       ylim=c(4,15)) +
  labs_pubr()
myo_gene_line = ggpar(myo_gene_line, xlab = "Day", ylab = "rlog-transformed Counts", legend="none", panel.labs.font="bold")
myo_gene_line

#ggexport(myo_gene_line, filename="./nosva_v4/Figures/myo_gene_line.tiff", width = 1200, height=1000, res=200)
```


```{r}
ggmaplot(resLFC_d0,
   fdr = 0.01, fc = 1, size = 0.4,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = as.vector(resLFC_d0$Symbol),
   legend = "top", top = 20,
   font.label = c("bold", 11), label.rectangle = TRUE,
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_minimal())
```


# KO vs WT

```{r}
dds_KOvsWT <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, design = ~ day + condition)
dds_KOvsWT = DESeq(dds_KOvsWT)
keep_KOvsWT <- rowSums(counts(dds_KOvsWT)) >= 1 # only keep genes that have at least one transcript across all samples
dds_KOvsWT <- dds_KOvsWT[keep_KOvsWT,]
resultsNames(dds_KOvsWT)
```

```{r}
rld_KOvsWT = rlog(dds_KOvsWT, blind=FALSE)
head(assay(rld_KOvsWT), 3)
```

```{r}
res_KOvsWT = results(dds_KOvsWT, name="condition_KO_vs_WT", pAdjustMethod = "fdr")
resLFC_KOvsWT = lfcShrink(dds_KOvsWT, res=res_KOvsWT, coef="condition_KO_vs_WT", type="ashr")
resLFC_KOvsWT$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_KOvsWT)), return.type="data.frame")$symbol
resLFC_KOvsWT = resLFC_KOvsWT[order(resLFC_KOvsWT$padj),]
head(resLFC_KOvsWT)
```

```{r}
KOvsWT_df = as.data.frame(resLFC_KOvsWT)
KOvsWT_df = subset(as.data.frame(KOvsWT_df[ROS_genes,]), select=c("Symbol", "log2FoldChange", "padj", "lfcSE"))
KOvsWT_df$log10padj = -log10(KOvsWT_df$padj)
KOvsWT_df$error_up = KOvsWT_df$log2FoldChange + KOvsWT_df$lfcSE
KOvsWT_df$error_down = KOvsWT_df$log2FoldChange - KOvsWT_df$lfcSE
KOvsWT_df = subset(KOvsWT_df, subset = padj<0.01)
KOvsWT_df

KOvsWT_bar = ggbarplot(KOvsWT_df, x="Symbol", y="log2FoldChange",
                    fill = "log10padj", color = "white",
                    sort.val = "asc", sort.by.groups = TRUE) + 
  gradient_fill("RdBu") +
  coord_flip() +
  scale_x_discrete(position = "left") +
  labs_pubr() +
  geom_errorbar(data=KOvsWT_df, aes(x=Symbol, ymin=error_down, ymax=error_up), width=0.4, size=1)

KOvsWT_bar = ggpar(KOvsWT_bar, legend = c(0.15,0.85), legend.title = "-log10(padj)", font.legend = c(14),
                font.xtickslab = 14, font.ytickslab = 14, 
                orientation = "horiz", xlab = FALSE, ylab = FALSE)

KOvsWT_bar
```

# Time-dependent differences
```{r}
dds_time <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, design = ~ day + condition + day:condition)
dds_time = DESeq(dds_time, test="LRT", reduced = ~ day + condition)
keep_time <- rowSums(counts(dds_time)) >= 1 # only keep genes that have at least one transcript across all samples
dds_time <- dds_time[keep_time,]
resultsNames(dds_time)
```

```{r}
rld_time = rlog(dds_time, blind=FALSE)
head(assay(rld_time), 3)
```

```{r}
res_time = results(dds_time, pAdjustMethod = "fdr")
head(res_time)
resLFC_time = lfcShrink(dds_time, coef="dayd.21.conditionKO", res=res_time, type="apeglm")
resLFC_time$Symbol = genes(edb, columns="symbol", filter=GeneIdFilter(rownames(resLFC_time)), return.type="data.frame")$symbol
resLFC_time = resLFC_time[order(resLFC_time$padj),]
head(resLFC_time[order(resLFC_time$padj),], 20)
```

```{r}
time_counts <- plotCounts(dds_time, which(resLFC_time$Symbol == "Myh3"), 
                   intgroup = c("day","condition"), returnData = TRUE)
head(time_counts)
ggplot(time_counts,
  aes(x = day, y = count, color = condition, group = condition)) + 
  geom_point() + stat_summary(fun.y=mean, geom="line") +
  scale_y_log10()
```

```{r}
topGenes <- head(order(resLFC_time$padj),20)
mat <- coef(dds_time)[topGenes, -c(1,2)]
thr <- 3 
mat[mat < -thr] <- -thr
mat[mat > thr] <- thr
pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
         cluster_col=FALSE)
```

